### PR TITLE
Add IPC shared memory example using the ipc.h API

### PR DESCRIPTION
**Pull request by ldorau**
### Description

Add IPC shared memory example using the ipc.h API.

### Checklist
- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

### CODE REVIEW

File: examples/basic/ipc_shm_ipcapi_producer.c [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+#include <umf/ipc.h>
+#include <umf/memory_pool.h>
+#include <umf/pools/pool_jemalloc.h>
+#include <umf/providers/provider_os_memory.h>
+
+#define INET_ADDR "127.0.0.1"
+#define MSG_SIZE 1024
+
+int main(int argc, char *argv[]) {
+    struct sockaddr_in consumer_addr;
```
Comment (vinser52) [inline examples/basic/ipc_shm_ipcapi_producer.c:24]: Sockets is a platform-specific way perhaps for simplification purposes we can use just files to communicate between processes. The producer creates an IPC handle and writes it to the file and the consumer process reads it from the file. Such example will be cross-platform.

File: examples/basic/ipc_shm_ipcapi.sh [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	fi
+	PTRACE_SCOPE_VAL=$(cat $PTRACE_SCOPE_FILE)
+	if [ $PTRACE_SCOPE_VAL -ne $VAL ]; then
+		echo "SKIP: setting ptrace_scope to 0 (classic ptrace permissions) FAILED - skipping the test"
+		exit 0
+	fi
+fi
+
+echo "Starting ipc_shm_ipcapi CONSUMER on port $PORT ..."
+./umf_example_ipc_shm_ipcapi_consumer $PORT &
```
Comment (vinser52) [inline examples/basic/ipc_shm_ipcapi.sh:35]: Instead of using a script to run the example can one process (e.g. consumer) just start the producer process?

Comment (ldorau) [reply examples/basic/ipc_shm_ipcapi.sh]: Such processes are not independent - the child process inherits some stuff from the parent process (AFAIR: environment, open file descriptors, ...). More over we have to set `ptrace_scope` to 0 - it is also done in this script.

File: examples/basic/ipc_shm_ipcapi_producer.c (hunk 1 above)
Comment (ldorau) [reply examples/basic/ipc_shm_ipcapi_producer.c:24]: I am afraid it would not be simpler. `recv()` waits for a message. We would have to implement a mechanism of waiting for a complete message being written into the file (checking if the message was written to the file and if it is complete).

Comment (bratpiorka) [reply examples/basic/ipc_shm_ipcapi_producer.c:24]: I think that since we already have this example working we could merge this version. As for now, we don't have plans to use IPC on Windows. OS provider IPC is also implemented on Linux only, on Windows it would return the NOT_SUPPORTED error for now.

File: examples/basic/ipc_shm_ipcapi.sh (hunk 2 above)
Comment (bratpiorka) [reply examples/basic/ipc_shm_ipcapi.sh]: I think that this way of running processes is closer to the requirements for UMF IPC functionality

File: examples/README.md
Comment (vinser52) [inline examples/README.md:1]: We should also update online documentation. If we plan to do it in a separate PR lets create an issue for that task.

File: examples/README.md [hunk 3]
```diff
@@ -38,3 +38,32 @@ and build this example Level Zero development package should be installed.
...
+- Consumer starts
+- Consumer creates a socket
+- Consumer listens for incoming connections
+- Producer starts
+- Producer's shared memory contains a number: 140722582213392
+- Producer gets the IPC handle
+- Producer creates a socket
+- Producer connects to the consumer
+- Consumer connects at IP 127.0.0.1 and port 47770 to the producer
+- Producer sents the IPC handle to the consumer (24 bytes)
```
Comment (vinser52) [inline examples/README.md:55]: `sents` -> `sends`

Comment (vinser52) [inline examples/README.md:55]: `(24 bytes)` is implementation details that might be changed. Let's omit such details.

File: examples/README.md [hunk 4]
```diff
@@ -38,3 +38,32 @@ and build this example Level Zero development package should be installed.
...
+- Consumer receives the IPC handle from the producer (24 bytes)
+- Consumer opens the IPC handle received from the producer
+- Consumer reads the number from the producer's shared memory: 140722582213392
+- Consumer writes a new number directly to the producer's shared memory: 70361291106696
+- Consumer sents a response message to the producer
```
Comment (vinser52) [inline examples/README.md:60]: `sents` -> `sends`

File: examples/basic/ipc_shm_ipcapi_consumer.c [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        goto err_destroy_OS_memory_provider;
+    }
+
+    producer_socket = consumer_connect(port);
+    if (producer_socket < 0) {
+        goto err_destroy_scalable_pool;
+    }
+
+    // receive the IPC handle from the producer's
+    ssize_t len = recv(producer_socket, recv_buffer, RECV_BUFF_SIZE, 0);
```
Comment (vinser52) [inline examples/basic/ipc_shm_ipcapi_consumer.c:132]: what if the size of the IPC handle is greater than `RECV_BUFF_SIZE`?

I think we need to send the size of the IPC handle first, and then receive the exact amount of bytes corresponding to IPC handle. It will make our example clearer and better demonstrate how to transfer IPC handle between processes.

File: examples/basic/ipc_shm_ipcapi_consumer.c [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+    fprintf(
+        stderr,
+        "[consumer] Received the IPC handle from the producer (%zi bytes)\n",
+        len);
+
+    void *SHM_ptr;
+    umf_result = umfOpenIPCHandle(scalable_pool, IPC_handle, &SHM_ptr);
+    if (umf_result == UMF_RESULT_ERROR_NOT_SUPPORTED) {
+        fprintf(stderr,
```
Comment (vinser52) [inline examples/basic/ipc_shm_ipcapi_consumer.c:148]: why do we need special processing for the `UMF_RESULT_ERROR_NOT_SUPPORTED` ? 
Furthermore, this block of code is not reachable, because if `umfOpenIPCHandle` is not supported then `umfGetIPCHandle` is not supported as well by the same memory provider on the producer side (the producer is not able to send the IPC handle in that case).

File: examples/basic/ipc_shm_ipcapi_producer.c [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                                         &OS_memory_provider);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        fprintf(stderr,
+                "[producer] ERROR: creating OS memory provider failed\n");
+        return -1;
+    }
+
+    size_t page_size;
+    umf_result =
+        umfMemoryProviderGetMinPageSize(OS_memory_provider, NULL, &page_size);
```
Comment (vinser52) [inline examples/basic/ipc_shm_ipcapi_producer.c:89]: Why do we need to use `page_size`? I think the allocation size does not matter for the IPC example.

File: examples/basic/ipc_shm_ipcapi_producer.c [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+    umf_result = umfPoolCreate(umfScalablePoolOps(), OS_memory_provider, NULL,
+                               0, &scalable_pool);
+    if (umf_result != UMF_RESULT_SUCCESS) {
+        fprintf(stderr,
+                "[producer] ERROR: creating the UMF jemalloc pool failed\n");
+        goto err_destroy_OS_memory_provider;
+    }
+
+    // Make 3 allocations of size: 1 page, 2 pages and 3 pages
+    void *ptr1, *ptr2, *IPC_shared_memory;
```
Comment (vinser52) [inline examples/basic/ipc_shm_ipcapi_producer.c:106]: for what reason we need `ptr1` and `ptr2`? It just blurs the attention.

**Review by vinser52 — CHANGES_REQUESTED** [review]

File: examples/basic/ipc_shm_ipcapi_consumer.c (hunk 6 above)
Comment (ldorau) [reply examples/basic/ipc_shm_ipcapi_consumer.c:148]: We need this to correctly handle the `UMF_RESULT_ERROR_NOT_SUPPORTED` error, because `umfOpenIPCHandle()` can return `UMF_RESULT_ERROR_NOT_SUPPORTED` when `! defined(__NR_pidfd_open) || ! defined(__NR_pidfd_getfd)`.
It is reachable on Ubuntu 20.04 where the `! defined(__NR_pidfd_open) || ! defined(__NR_pidfd_getfd)` condition is true:
https://github.com/ldorau/unified-memory-framework/actions/runs/9002739451/job/24731880245

Comment (vinser52) [reply examples/basic/ipc_shm_ipcapi_consumer.c:148]: Ah, I see.

File: examples/basic/ipc_shm_ipcapi_producer.c (hunk 7 above)
Comment (vinser52) [reply examples/basic/ipc_shm_ipcapi_producer.c:89]: Now I understand the reason. The `pool_proxy` implementation does not automatically round up the size to be multiple of the page size. But as I described in #475, I think the `pool_proxy` should be extended to support arbitrary allocation sizes.

File: examples/README.md (hunk 3 above)
Comment (ldorau) [reply examples/README.md]: Done

Comment (ldorau) [reply examples/README.md]: Done

File: examples/README.md (hunk 4 above)
Comment (ldorau) [reply examples/README.md]: Done

File: examples/basic/ipc_shm_ipcapi_consumer.c (hunk 5 above)
Comment (ldorau) [reply examples/basic/ipc_shm_ipcapi_consumer.c:135]: Done

File: examples/basic/ipc_shm_ipcapi_producer.c (hunk 7 above)
Comment (ldorau) [reply examples/basic/ipc_shm_ipcapi_producer.c]: Removed

File: examples/basic/ipc_shm_ipcapi_producer.c (hunk 8 above)
Comment (ldorau) [reply examples/basic/ipc_shm_ipcapi_producer.c]: It was a test for correctness of `fd_offset` - it is done in the IPC test.
Removed.

File: examples/basic/ipc_shm_ipcapi_producer.c (hunk 7 above)
Comment (ldorau) [reply examples/basic/ipc_shm_ipcapi_producer.c]: A size can be any now, because we use `umfPoolCalloc(scalable_pool)` to allocate this memory.

File: examples/README.md
Comment (ldorau) [reply examples/README.md:1]: Done: https://github.com/oneapi-src/unified-memory-framework/issues/477

**Review by vinser52 — APPROVED** [review]

**Review by bratpiorka — APPROVED** [review]

**Comment by bratpiorka** [comment]
@vinser52 please close all conversations if you think the changes are ok :)

**Comment by vinser52** [comment]
> @vinser52 please close all conversations if you think the changes are ok :)

Done
